doc/cabal-package.rst

```diff
@@ -1017,24 +1017,70 @@ disambiguation purposes. Example:
     $ cabal repl bench:baz
 
 Freezing dependency versions
-""""""""""""""""""""""""""""
-
-If a package is built in several different environments, such as a
-development environment, a staging environment and a production
-environment, it may be necessary or desirable to ensure that the same
-dependency versions are selected in each environment. This can be done
-with the ``freeze`` command:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 .. code-block:: console
 
     $ cabal freeze
 
-The command writes the selected version for all dependencies to the
-``cabal.config`` file. All environments which share this file will use
-the dependency versions specified in it.
+generates ``cabal.project.freeze`` file, which describes the exact dependency tree as it was resolved at that moment by Cabal.
+This means it captures an exact version of every dependency, including dependencies of dependencies, recursively all the way.
+
+Since ``cabal`` reads ``cabal.project.freeze`` when present, and takes into consideration the version constraints in it,
+this means that by producing ``cabal.project.freeze`` you are guaranteed that every future ``cabal`` call will use the exact same set of dependencies,
+regardless of any updates (even patches) that might get published for these dependencies in the meantime.
+Therefore, we have effectively "frozen" the dependencies in place, making our build consistent and reproducible.
+
+``cabal.project.freeze`` is intended to be committed to the version control.
```

Comment:
I think we should warn about the fact `cabal.project.freeze` is heavily os dependent, spcially if you are using native libs.
The thing is even harder if you support windows, as widely used libs have conditionals in the .cabal file using `os(windows)`
See the caveats in a recent issue: https://github.com/haskell/cabal/issues/8059

Reply:
An alternative could be only frreze the hackage index state, as commented here: https://github.com/haskell/cabal/issues/8059#issuecomment-1076024522

Reply:
Also i would note it also freezes cabal flags and it could drive to unexpected results: https://github.com/haskell/cabal/issues/7944

Reply:
Ah, and it is also dependent on ghc version :slightly_smiling_face: : #7367

Reply:
And there is another request about makes docs clear on what to do when you have both a executable and a library: #5750.

Reply:
@Martinsos sorry, i dont want to scare you. I am reviewing the existing issues about freeze and it turns out it has several caveats.
The pr looks nice overall i would ask only to mention the fact they are (or they might be) dependent on ghc and os, rest of things (freeze index state alternative, flags and #5750) could be added in other prs

Reply:
Hah @jneira, no worries, nothing to be scared of :D, I am glad you are sharing all this info as I wasn't aware of it!
Since we are using `cabal freeze` in our project currently, I am anyway trying to learn as much as I can about it and how to use it correctly (as is visible in #8059 and #8047), and while doing that I discovered some of these caveats on my own, while the rest I am happy to learn about.
This PR is really a way for me to capture all that I learn on this "journey" and make it a bit easier for others, therefore I am happy to capture as much info as we can.

Let me play a bit more with `freeze` following days, and I will be coming back here to either update PR with what I learn, or ask additional questions, until we have something that gives a good overview of `freeze`.

Some questions immediately:
1. flags -> do I want to freeze them or not? I have to admit I don't understand this topic well enough -> what would be your quick take on it? Are there use cases where I want to freeze them, are there ones where i don't want to? What are those? Is there a general quick advice on how to think about this?
2. dependent on ghc version -> makes sense, it is not a problem for the project I am currently working on as it is executable that we build with specific ghc always, but I guess it can be a problem for some libraries? Although I thought for them it is not recommended to use freeze file. I guess I am asking what is the situation where one has multi-ghc project and needs freeze file. Probaly should ask that on the relevant issue though and not here -> i is not so important right now anyway.
3. As for using freeze when project has both executable and a library -> do we know the answer? I am not sure -> I guess it comes down to ensuring that freeze file doesn't get included into the sdist, therefore it doesn't get distributed with a library, but it does get distributed with an executable?

Btw, looking at different issues -> different GHC versions, different OS-es, it seems to me the proper solution might be in the direction of having per-environment (where environment is defined as combo of os, ghc and arch) freeze files. Meaning that user could define environments it cares about, and then on `cabal freeze`, freeze file for each of those environments would be generated (is that possible? Can we generate freeze file for osx while on linux -> I guess in theory we should be able to, right?). Additionally, cabal would use correct freeze file based on the current environment.
But I will take this question to #8059 .

Reply:
> flags -> do I want to freeze them or not? I have to admit I don't understand this topic well enough -> what would be your quick take on it? Are there use cases where I want to freeze them, are there ones where i don't want to? What are those? Is there a general quick advice on how to think about this?

well there are two kind of cabal flags, automatic and manual and that determines its freezing imo:
* the cabal solver tries the automatic flags to get a working build plan, so freeze them makes more sense imo. The user usually dont care about them (if they work :-) ) although you can set them explicitly. The archetipical example is `integer-gmp`/`integer-simple` which switch between two implementations of the same api.
* manual ones must be explicitly set by the user, you can use the cli `-f` option or store it in the cabal.project(.local) if you want to avoid set them in every invocation. That is the case of #7944 where you have a `debug` flag you only want to set sometimes. So freeze them could make sense or could not, depending of its usage.

But `cabal freeze` store them all so you have to delete manually the unwanted ones